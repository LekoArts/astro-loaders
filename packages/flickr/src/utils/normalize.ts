import type { z } from 'astro/zod'
import type { NormalizedPhoto } from '../schema.js'
import type { FlickrPhoto } from '../types/flickr'
import { SIZES } from '../constants.js'

type Photo = z.infer<typeof NormalizedPhoto>
type SizesArray = Array<keyof typeof SIZES>
type SizesUnion = keyof typeof SIZES

interface ImageUrl {
  url: string
  height: number
  width: number
  orientation: 'landscape' | 'portrait' | 'square'
}

/**
 * The responses you get back from Flickr's API are a mess. They mix naming conventions, different data types for the same category, and so on. This function tries to normalize the data a bit.
 */
export function normalizePhoto(photo: FlickrPhoto): Photo {
  const { server, farm, secret, ...rest } = photo
  const copy = { ...rest } as Partial<FlickrPhoto>
  const output = {
    id: rest.id,
    owner: rest.owner,
    title: rest.title,
  } as Photo

  ;(Object.keys(SIZES) as SizesArray).forEach((suffix) => {
    if (Object.prototype.hasOwnProperty.call(copy, `height_${suffix}`)) {
      copy[`height_${suffix}`] = Number.parseInt(copy[`height_${suffix}`] as string, 10)
    }
    if (Object.prototype.hasOwnProperty.call(copy, `width_${suffix}`)) {
      copy[`width_${suffix}`] = Number.parseInt(copy[`width_${suffix}`] as string, 10)
    }
  })

  if (Object.prototype.hasOwnProperty.call(copy, 'ispublic')) {
    output.is_public = copy.ispublic === 1
  }

  if (Object.prototype.hasOwnProperty.call(copy, 'isfriend')) {
    output.is_friend = copy.isfriend === 1
  }

  if (Object.prototype.hasOwnProperty.call(copy, 'isfamily')) {
    output.is_family = copy.isfamily === 1
  }

  if (Object.prototype.hasOwnProperty.call(copy, 'description')) {
    output.description = copy.description?._content
  }

  if (Object.prototype.hasOwnProperty.call(copy, 'lastupdate')) {
    if (copy.lastupdate) {
      output.date_last_update = new Date(+copy.lastupdate * 1000)
    }
  }

  if (Object.prototype.hasOwnProperty.call(copy, 'datetaken')) {
    if (copy.datetaken) {
      output.date_taken = new Date(copy.datetaken)
    }
  }

  if (Object.prototype.hasOwnProperty.call(copy, 'views')) {
    if (copy.views) {
      output.views = Number.parseInt(copy.views, 10)
    }
  }

  if (Object.prototype.hasOwnProperty.call(copy, 'media')) {
    output.media = copy.media
  }

  if (Object.prototype.hasOwnProperty.call(copy, 'media_status')) {
    output.media_status = copy.media_status
  }

  output.imageUrls = {}

  for (const key in copy) {
    if (Object.prototype.hasOwnProperty.call(copy, key)) {
      const firstElem = key.toString().split(`_`).shift()
      const lastElem = key.toString().split(`_`).pop() as SizesUnion

      if (firstElem && lastElem && (Object.keys(SIZES) as Array<keyof typeof SIZES>).includes(lastElem)) {
        const sizeKey = SIZES[lastElem]
        // @ts-expect-error - Fixme
        const newElem = firstElem === `url` ? { [firstElem]: copy[key] } : { [firstElem]: copy[key] }
        // @ts-expect-error - Fixme
        output.imageUrls[sizeKey] = { ...output.imageUrls[sizeKey], ...newElem }
      }

      for (const image in output.imageUrls) {
        if (Object.prototype.hasOwnProperty.call(output.imageUrls, image)) {
          // @ts-expect-error - Fixme
          const element: ImageUrl = output.imageUrls[image]

          element.orientation = element.width === element.height ? `square` : element.width > element.height ? `landscape` : `portrait`
        }
      }
    }
  }

  return output
}
