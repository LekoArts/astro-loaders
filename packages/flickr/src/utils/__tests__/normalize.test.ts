import { describe, expect, it } from 'vitest'
import { normalizePhoto } from '../normalize'

import peopleGetPhotosFixture from './fixtures/people-getPhotos.json'

describe('normalizePhoto', () => {
  it('should normalize people-getPhotos response', () => {
    const normalized = normalizePhoto(peopleGetPhotosFixture)

    expect(normalized).toMatchSnapshot()
  })
})
