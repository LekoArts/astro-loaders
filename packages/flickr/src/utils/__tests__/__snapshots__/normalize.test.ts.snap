// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`normalizePhoto > should normalize people-getPhotos response 1`] = `
{
  "date_last_update": 2023-08-24T14:22:22.000Z,
  "date_taken": 2023-07-20T19:04:31.000Z,
  "description": "Photo",
  "id": "53077440094",
  "imageUrls": {
    "100px": {
      "height": 56,
      "orientation": "landscape",
      "url": "URL",
      "width": 100,
    },
    "1024px": {
      "height": 576,
      "orientation": "landscape",
      "url": "URL",
      "width": 1024,
    },
    "1600px": {
      "height": 576,
      "orientation": "landscape",
      "url": "URL",
      "width": 1024,
    },
    "2048px": {
      "height": 1024,
      "orientation": "square",
      "url": "URL",
      "width": 1024,
    },
    "240px": {
      "height": 135,
      "orientation": "landscape",
      "url": "URL",
      "width": 240,
    },
    "320px": {
      "height": 180,
      "orientation": "landscape",
      "url": "URL",
      "width": 320,
    },
    "500px": {
      "height": 281,
      "orientation": "landscape",
      "url": "URL",
      "width": 500,
    },
    "640px": {
      "height": 360,
      "orientation": "landscape",
      "url": "URL",
      "width": 640,
    },
    "800px": {
      "height": 450,
      "orientation": "landscape",
      "url": "URL",
      "width": 800,
    },
    "original": {
      "height": 3000,
      "orientation": "portrait",
      "url": "URL",
      "width": 1024,
    },
    "sq_150px": {
      "height": 150,
      "orientation": "square",
      "url": "URL",
      "width": 150,
    },
    "sq_75px": {
      "height": 75,
      "orientation": "square",
      "url": "URL",
      "width": 75,
    },
  },
  "is_family": false,
  "is_friend": false,
  "is_public": true,
  "media": "photo",
  "media_status": "ready",
  "owner": "192975453@N04",
  "title": "DSCF1603",
  "views": 24,
}
`;
